File: apps/siakad-tu/src/domain/student/Student.js
// apps/siakad-tu/src/domain/student/Student.js

import { NISN } from './value-objects/NISN.js'
import { NIS } from './value-objects/NIS.js'
import { FullName } from './value-objects/FullName.js'
import { BirthDate } from './value-objects/BirthDate.js'
import { Gender } from './value-objects/Gender.js'
import { Address } from './value-objects/Address.js'
import { ContactInfo } from './value-objects/ContactInfo.js'
import { GuardianInfo } from './value-objects/GuardianInfo.js'
import { StudentStatus } from './value-objects/StudentStatus.js'
import { Enrollment } from './Enrollment.js'
import {
  StudentRegistered,
  StudentEnrolled,
  StudentTransferred,
  StudentGraduated,
  StudentProfileUpdated,
} from './events/index.js' // index.js yang export semua events

/**
 * Aggregate Root: Student
 *
 * Entry point untuk semua perubahan pada data siswa.
 * Menjaga invariants, menghasilkan domain events.
 *
 * INVARIANTS:
 * 1. NISN unik dan valid (10 digit) — dijaga oleh Value Object
 * 2. NIS unik dalam satu sekolah — dijaga oleh Repository
 * 3. Siswa tidak bisa di-enroll 2x di periode yang sama
 * 4. Siswa "graduated/transferred/dropped" tidak bisa di-enroll lagi
 * 5. Siswa "active" harus punya minimal 1 enrollment aktif
 * 6. Enrollment terikat pada OperationalContext (schoolId + periodId)
 */
export class Student {
  #studentId
  #schoolId // dari OperationalContext saat registrasi
  #nisn
  #nis
  #fullName
  #birthDate
  #gender
  #address
  #contactInfo
  #guardianInfo
  #status
  #enrollments // Entity[]
  #events // DomainEvent[]

  constructor({
    studentId,
    schoolId,
    nisn,
    nis,
    fullName,
    birthDate,
    gender,
    address,
    contactInfo,
    guardianInfo,
    status = StudentStatus.ACTIVE,
    enrollments = [],
  }) {
    if (!studentId) throw new Error('Student: studentId wajib.')
    if (!schoolId) throw new Error('Student: schoolId wajib (dari OperationalContext).')

    this.#studentId = studentId
    this.#schoolId = schoolId
    this.#nisn = nisn instanceof NISN ? nisn : new NISN(nisn)
    this.#nis = nis instanceof NIS ? nis : new NIS(nis)
    this.#fullName =
      fullName instanceof FullName
        ? fullName
        : new FullName(fullName.firstName, fullName.middleName, fullName.lastName)
    this.#birthDate = birthDate instanceof BirthDate ? birthDate : new BirthDate(birthDate)
    this.#gender = gender instanceof Gender ? gender : gender
    this.#address = address instanceof Address ? address : new Address(address)
    this.#contactInfo =
      contactInfo instanceof ContactInfo ? contactInfo : new ContactInfo(contactInfo)
    this.#guardianInfo =
      guardianInfo instanceof GuardianInfo ? guardianInfo : new GuardianInfo(guardianInfo)
    this.#status = status instanceof StudentStatus ? status : status
    this.#enrollments = enrollments.map((e) =>
      e instanceof Enrollment ? e : Enrollment.fromJSON(e),
    )
    this.#events = []
  }

  // === GETTERS ===
  get studentId() {
    return this.#studentId
  }
  get schoolId() {
    return this.#schoolId
  }
  get nisn() {
    return this.#nisn
  }
  get nis() {
    return this.#nis
  }
  get fullName() {
    return this.#fullName
  }
  get birthDate() {
    return this.#birthDate
  }
  get gender() {
    return this.#gender
  }
  get address() {
    return this.#address
  }
  get contactInfo() {
    return this.#contactInfo
  }
  get guardianInfo() {
    return this.#guardianInfo
  }
  get status() {
    return this.#status
  }
  get enrollments() {
    return [...this.#enrollments]
  }

  // === BEHAVIORS ===

  /**
   * Register siswa baru
   * Dipanggil saat siswa pertama kali didaftarkan di sekolah
   */
  static register({
    studentId,
    schoolId,
    nisn,
    nis,
    fullName,
    birthDate,
    gender,
    address,
    contactInfo,
    guardianInfo,
  }) {
    const student = new Student({
      studentId,
      schoolId,
      nisn,
      nis,
      fullName,
      birthDate,
      gender,
      address,
      contactInfo,
      guardianInfo,
      status: StudentStatus.ACTIVE,
    })

    // Record event
    student.#events.push(
      new StudentRegistered(studentId, schoolId, {
        nisn: student.#nisn.value,
        nis: student.#nis.value,
        fullName: student.#fullName.displayName,
      }),
    )

    return student
  }

  /**
   * Enroll siswa ke kelas pada periode tertentu
   *
   * INVARIANTS:
   * - Status harus ACTIVE
   * - Tidak boleh ada enrollment aktif di periode yang sama
   */
  enroll({ enrollmentId, periodId, classId, enrollmentDate }) {
    // Invariant 4: Hanya ACTIVE yang bisa di-enroll
    if (!this.#status.canBeEnrolled()) {
      throw new Error(`Siswa dengan status "${this.#status.label}" tidak bisa di-enroll.`)
    }

    // Invariant 3: Tidak boleh ada enrollment aktif di periode yang sama
    const existingActive = this.#enrollments.find((e) => e.isForPeriod(periodId) && e.isActive())
    if (existingActive) {
      throw new Error(`Siswa sudah memiliki enrollment aktif di periode ${periodId}.`)
    }

    const enrollment = new Enrollment({
      enrollmentId,
      schoolId: this.#schoolId,
      periodId,
      classId,
      enrollmentDate,
    })

    this.#enrollments.push(enrollment)

    // Record event
    this.#events.push(
      new StudentEnrolled(this.#studentId, this.#schoolId, {
        enrollmentId,
        periodId,
        classId,
      }),
    )
  }

  /**
   * Update profil siswa (address, contact, guardian)
   */
  updateProfile({ address, contactInfo, guardianInfo }) {
    if (address) {
      this.#address = address instanceof Address ? address : new Address(address)
    }
    if (contactInfo) {
      this.#contactInfo =
        contactInfo instanceof ContactInfo ? contactInfo : new ContactInfo(contactInfo)
    }
    if (guardianInfo) {
      this.#guardianInfo =
        guardianInfo instanceof GuardianInfo ? guardianInfo : new GuardianInfo(guardianInfo)
    }

    this.#events.push(
      new StudentProfileUpdated(this.#studentId, this.#schoolId, {
        updatedFields: Object.keys({ address, contactInfo, guardianInfo }).filter(
          (k) =>
            [address, contactInfo, guardianInfo][
              ['address', 'contactInfo', 'guardianInfo'].indexOf(k)
            ],
        ),
      }),
    )
  }

  /**
   * Tandai siswa sebagai lulus
   */
  graduate() {
    if (!this.#status.isActive()) {
      throw new Error(`Siswa dengan status "${this.#status.label}" tidak bisa diluluskan.`)
    }

    // Complete semua enrollment aktif
    this.#enrollments.filter((e) => e.isActive()).forEach((e) => e.complete())

    this.#status = StudentStatus.GRADUATED

    this.#events.push(
      new StudentGraduated(this.#studentId, this.#schoolId, {
        graduationDate: new Date().toISOString(),
      }),
    )
  }

  /**
   * Tandai siswa sebagai pindah
   */
  transfer({ reason, targetSchool }) {
    if (!this.#status.isActive()) {
      throw new Error(`Siswa dengan status "${this.#status.label}" tidak bisa dipindahkan.`)
    }

    this.#enrollments.filter((e) => e.isActive()).forEach((e) => e.complete())

    this.#status = StudentStatus.TRANSFERRED

    this.#events.push(
      new StudentTransferred(this.#studentId, this.#schoolId, {
        reason,
        targetSchool,
        transferDate: new Date().toISOString(),
      }),
    )
  }

  /**
   * Cek apakah siswa memiliki enrollment aktif di periode tertentu
   */
  hasActiveEnrollmentInPeriod(periodId) {
    return this.#enrollments.some((e) => e.isForPeriod(periodId) && e.isActive())
  }

  /**
   * Dapatkan enrollment aktif saat ini (terbaru)
   */
  getCurrentEnrollment() {
    return (
      this.#enrollments
        .filter((e) => e.isActive())
        .sort((a, b) => b.enrollmentDate - a.enrollmentDate)[0] || null
    )
  }

  // === EVENT MANAGEMENT ===

  getUncommittedEvents() {
    return [...this.#events]
  }

  clearEvents() {
    this.#events = []
  }

  // === SERIALIZATION ===

  toJSON() {
    return {
      studentId: this.#studentId,
      schoolId: this.#schoolId,
      nisn: this.#nisn.toJSON(),
      nis: this.#nis.toJSON(),
      fullName: this.#fullName.toJSON(),
      birthDate: this.#birthDate.toJSON(),
      gender: this.#gender.toJSON(),
      address: this.#address.toJSON(),
      contactInfo: this.#contactInfo.toJSON(),
      guardianInfo: this.#guardianInfo.toJSON(),
      status: this.#status.toJSON(),
      enrollments: this.#enrollments.map((e) => e.toJSON()),
    }
  }

  static fromJSON(data) {
    return new Student({
      studentId: data.studentId,
      schoolId: data.schoolId,
      nisn: NISN.fromJSON(data.nisn),
      nis: NIS.fromJSON(data.nis),
      fullName: FullName.fromJSON(data.fullName),
      birthDate: BirthDate.fromJSON(data.birthDate),
      gender: Gender.fromJSON(data.gender),
      address: Address.fromJSON(data.address),
      contactInfo: ContactInfo.fromJSON(data.contactInfo),
      guardianInfo: GuardianInfo.fromJSON(data.guardianInfo),
      status: StudentStatus.fromJSON(data.status),
      enrollments: data.enrollments || [],
    })
  }
}

File: services/tu-core/internal/domain/student/models/student.go
package models

import (
	"errors"
	"time"

	"sekolah-platform/platform/context"
	"sekolah-platform/platform/events/types"
	"sekolah-platform/services/tu-core/internal/domain/student/events"
	"sekolah-platform/services/tu-core/internal/domain/student/models/valueobjects"

	"github.com/google/uuid"
)

// ============================================================================
// VALUE OBJECTS (Local Structs untuk Data Transfer)
// ============================================================================

// Address adalah struct untuk data alamat.
type Address struct {
	Street     string
	RtRw       string
	Village    string
	District   string
	City       string
	PostalCode string
}

// ContactInfo adalah struct untuk data kontak.
type ContactInfo struct {
	Phone string
	Email string
}

// GuardianInfo adalah struct untuk data wali/orang tua.
type GuardianInfo struct {
	Name       string
	Relation   string // father, mother, guardian
	Phone      string
	Occupation string
}

// ============================================================================
// AGGREGATE ROOT
// ============================================================================

// Student adalah Aggregate Root untuk domain Student.
type Student struct {
	studentID    string
	schoolID     string
	nisn         valueobjects.NISN
	nis          valueobjects.NIS
	fullName     valueobjects.FullName
	birthDate    time.Time
	gender       valueobjects.Gender
	address      Address
	contactInfo  ContactInfo
	guardianInfo GuardianInfo
	status       valueobjects.StudentStatus
	enrollments  []*Enrollment
	events       []types.DomainEvent
	createdAt    time.Time
	updatedAt    time.Time
}

// StudentData adalah data input untuk membuat Student.
type StudentData struct {
	StudentID    string
	SchoolID     string
	NISN         string
	NIS          string
	FirstName    string
	MiddleName   string
	LastName     string
	BirthDate    time.Time
	Gender       string
	Address      Address
	ContactInfo  ContactInfo
	GuardianInfo GuardianInfo
}

// NewStudent membuat Student baru.
func NewStudent(data StudentData, ctx *context.OperationalContext) (*Student, error) {
	if data.StudentID == "" {
		data.StudentID = uuid.New().String()
	}

	nisn, err := valueobjects.NewNISN(data.NISN)
	if err != nil {
		return nil, err
	}

	nis, err := valueobjects.NewNIS(data.NIS)
	if err != nil {
		return nil, err
	}

	fullName, err := valueobjects.NewFullName(data.FirstName, data.MiddleName, data.LastName)
	if err != nil {
		return nil, err
	}

	gender, err := valueobjects.NewGender(data.Gender)
	if err != nil {
		return nil, err
	}

	student := &Student{
		studentID:    data.StudentID,
		schoolID:     ctx.SchoolID,
		nisn:         nisn,
		nis:          nis,
		fullName:     fullName,
		birthDate:    data.BirthDate,
		gender:       gender,
		address:      data.Address,
		contactInfo:  data.ContactInfo,
		guardianInfo: data.GuardianInfo,
		status:       valueobjects.StudentStatusActive,
		enrollments:  []*Enrollment{},
		events:       []types.DomainEvent{},
		createdAt:    time.Now().UTC(),
		updatedAt:    time.Now().UTC(),
	}

	// ✅ AMAN: Mengirim Payload Struct, BUKAN pointer ke struct Student
	student.recordEvent(events.NewStudentCreated(events.StudentCreatedPayload{
		StudentID: student.studentID,
		NISN:      student.nisn.Value(),
		NIS:       student.nis.Value(),
		FullName:  student.fullName.DisplayName(),
		Gender:    student.gender.Code(),
		BirthDate: student.birthDate,
	}, ctx))

	return student, nil
}

// === GETTERS ===
func (s *Student) StudentID() string                  { return s.studentID }
func (s *Student) SchoolID() string                   { return s.schoolID }
func (s *Student) NISN() valueobjects.NISN            { return s.nisn }
func (s *Student) NIS() valueobjects.NIS              { return s.nis }
func (s *Student) FullName() valueobjects.FullName    { return s.fullName }
func (s *Student) BirthDate() time.Time               { return s.birthDate }
func (s *Student) Gender() valueobjects.Gender        { return s.gender }
func (s *Student) Address() Address                   { return s.address }
func (s *Student) ContactInfo() ContactInfo           { return s.contactInfo }
func (s *Student) GuardianInfo() GuardianInfo         { return s.guardianInfo }
func (s *Student) Status() valueobjects.StudentStatus { return s.status }
func (s *Student) Enrollments() []*Enrollment         { return s.enrollments }
func (s *Student) CreatedAt() time.Time               { return s.createdAt }
func (s *Student) UpdatedAt() time.Time               { return s.updatedAt }

// === BEHAVIORS ===

func (s *Student) Enroll(enrollmentID, periodID, classID string, ctx *context.OperationalContext) error {
	if !s.status.CanBeEnrolled() {
		return errors.New("siswa dengan status " + s.status.Label() + " tidak bisa di-enroll")
	}

	for _, e := range s.enrollments {
		if e.PeriodID() == periodID && e.IsActive() {
			return errors.New("siswa sudah memiliki enrollment aktif di periode " + periodID)
		}
	}

	enrollment := NewEnrollment(enrollmentID, s.schoolID, periodID, classID)
	s.enrollments = append(s.enrollments, enrollment)
	s.updatedAt = time.Now().UTC()

	// ✅ AMAN: Mengirim Payload Struct
	s.recordEvent(events.NewStudentEnrolled(events.StudentEnrolledPayload{
		StudentID:    s.studentID,
		EnrollmentID: enrollment.EnrollmentID(),
		PeriodID:     enrollment.PeriodID(),
		ClassID:      enrollment.ClassID(),
	}, ctx))

	return nil
}

func (s *Student) Graduate(ctx *context.OperationalContext) error {
	if !s.status.IsActive() {
		return errors.New("siswa dengan status " + s.status.Label() + " tidak bisa diluluskan")
	}

	for _, e := range s.enrollments {
		if e.IsActive() {
			e.Complete()
		}
	}

	s.status = valueobjects.StudentStatusGraduated
	s.updatedAt = time.Now().UTC()

	// ✅ AMAN: Mengirim Payload Struct
	s.recordEvent(events.NewStudentGraduated(events.StudentGraduatedPayload{
		StudentID:      s.studentID,
		GraduationDate: time.Now().UTC(),
	}, ctx))

	return nil
}

func (s *Student) Transfer(targetSchool, reason string, ctx *context.OperationalContext) error {
	if !s.status.IsActive() {
		return errors.New("siswa dengan status " + s.status.Label() + " tidak bisa dipindahkan")
	}

	for _, e := range s.enrollments {
		if e.IsActive() {
			e.Complete()
		}
	}

	s.status = valueobjects.StudentStatusTransferred
	s.updatedAt = time.Now().UTC()

	// ✅ AMAN: Mengirim Payload Struct
	s.recordEvent(events.NewStudentTransferred(events.StudentTransferredPayload{
		StudentID:    s.studentID,
		TargetSchool: targetSchool,
		Reason:       reason,
		TransferDate: time.Now().UTC(),
	}, ctx))

	return nil
}

// === EVENT MANAGEMENT ===

func (s *Student) recordEvent(event types.DomainEvent) {
	s.events = append(s.events, event)
}

func (s *Student) UncommittedEvents() []types.DomainEvent {
	return s.events
}

func (s *Student) ClearEvents() {
	s.events = []types.DomainEvent{}
}
